src/point_one/fusion_engine/parsers/fusion_engine_framer.h

```diff
@@ -0,0 +1,168 @@
+/**************************************************************************/ /**
+ * @brief FusionEngine message framer.
+ * @file
+ ******************************************************************************/
+
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <memory>
+
+#include "point_one/fusion_engine/messages/defs.h"
+
+namespace point_one {
+namespace fusion_engine {
+namespace parsers {
+
+/**
+ * @brief Frame and validate incoming FusionEngine messages.
+ *
+ * This class locates and validates FusionEngine messages within a stream of
+ * binary data. Data may be stored in an internally allocated buffer, or in an
+ * external buffer supplied by the user.
+ *
+ * The callback function provided to @ref SetMessageCallback() will be called
+ * each time a complete message is received. Any messages that do not pass the
+ * CRC check, or that are too big to be stored in the data buffer, will be
+ * discarded.
+ *
+ * Example usage:
+ * ```cpp
+ * void MessageReceived(const MessageHeader& header, const void* payload) {
+ *   if (header.message_type == MessageType::POSE) {
+ *     auto& contents = *static_cast<const PoseMessage*>(payload);
+ *     ...
+ *   }
+ * }
+ *
+ * FusionEngineFramer framer(1024);
+ * framer.SetMessageCallback(MessageReceived);
+ * framer.OnData(my_data, my_data_size);
+ * ```
+ */
+class FusionEngineFramer {
+public:
+  /**
+   * @brief Construct a framer instance with no buffer allocated.
+   *
+   * @note
+   * You must call @ref SetBuffer() to assign a buffer, otherwise all incoming
+   * data will be discarded.
+   */
+  FusionEngineFramer() = default;
+
+  /**
+   * @brief Construct a framer instance with an internally allocated buffer.
+   *
+   * @param capacity_bytes The maximum framing buffer capacity (in bytes).
+   */
+  explicit FusionEngineFramer(size_t capacity_bytes)
+      : FusionEngineFramer(nullptr, capacity_bytes) {}
+
+  /**
+   * @brief Construct a framer instance with a user-specified buffer.
+   *
+   * @post
+   * `buffer` must exist for the lifetime of this instance.
+   *
+   * @param buffer The framing buffer to use. Set to `nullptr` to allocate a
+   *        buffer internally.
+   * @param capacity_bytes The maximum framing buffer capacity (in bytes).
+   */
+  FusionEngineFramer(void* buffer, size_t capacity_bytes);
+
+  /**
+   * @brief Set the buffer to use for message framing.
+   *
+   * @post
+   * `buffer` must exist for the lifetime of this instance.
+   *
+   * @param buffer The framing buffer to use. Set to `nullptr` to allocate a
+   *        buffer internally.
+   * @param capacity_bytes The maximum framing buffer capacity (in bytes).
+   */
+  void SetBuffer(void* buffer, size_t capacity_bytes);
+
+  /**
+   * @brief Enable/disable warnings for CRC and "message too large" failures.
```

Comment:
I think it's worth giving a brief explanation of a case where this is use (i.e., if the data stream is expected to have non-FusionEngine data, these failures are expected. )